src/off_chain/historical_data/cow_endpoint_surplus.py

```diff
@@ -0,0 +1,295 @@
+    Adds all hashes to a list (between blocks or single hash) and fetches competition endpoint data for all of the hashes.
+    Then calls on the get_order_surplus in a loop to compute potential better surplus for each settlement, if any.
+    """
+
+    def get_surplus_by_input(
+        self,
+        start_block: Optional[int] = None,
+        end_block: Optional[int] = None,
+        settlement_hash: Optional[str] = None,
+    ) -> None:
+        settlement_hashes_list = []
+
+        if settlement_hash is not None:
+            settlement_hashes_list.append(settlement_hash)
+
+        elif start_block is not None and end_block is not None:
+            # Etherscan endpoint call for settlements between start and end block
+            settlement_hashes_list = self.get_settlement_hashes(start_block, end_block)
+        solver_competition_data = self.get_solver_competition_data(
+            settlement_hashes_list
+        )
+        for comp_data in solver_competition_data:
+            self.get_order_surplus(comp_data)
+
+    """
+    This function gets all hashes for a contract address between two blocks
+    """
+
+    def get_settlement_hashes(self, start_block: int, end_block: int) -> List[str]:
+        etherscan_url = f"https://api.etherscan.io/api?module=account&action=txlist&address=0x9008D19f58AAbD9eD0D60971565AA8510560ab41&startblock={start_block}&endblock={end_block}&sort=desc&apikey={directory.ETHERSCAN_KEY}"
+        # all "result" go into results (based on API return value names from docs)
+        try:
+            settlements = json.loads(
+                (
+                    requests.get(
+                        etherscan_url,
+                        headers=header,
+                        timeout=1000000,
+                    )
+                ).text
+            )["result"]
+            settlement_hashes_list = []
+            for settlement in settlements:
+                settlement_hashes_list.append(settlement["hash"])
+        except ValueError:
+            self.logger.critical("etherscan error.")
+
+        return settlement_hashes_list
+
+    """
+    This function uses a list of tx hashes to fetch and assemble competition data
+    for each of the tx hashes and returns it to get_surplus_by_input for further surplus calculation.
+    """
+
+    def get_solver_competition_data(
+        self, settlement_hashes_list: List[str]
+    ) -> List[Dict[str, Any]]:
+        solver_competition_data = []
+        for tx_hash in settlement_hashes_list:
+            endpoint_url = f"https://api.cow.fi/mainnet/api/v1/solver_competition/by_tx_hash/{tx_hash}"
+            json_competition_data = requests.get(
+                endpoint_url,
+                headers=header,
+                timeout=1000000,
+            )
+            if json_competition_data.status_code == 200:
+                solver_competition_data.append(json.loads(json_competition_data.text))
+
+        return solver_competition_data
```

Comment:
I had this in mind initially. However after testing over vast ranges, it seems that etherscan api picks up failed transactions as well for the cow settlement address. For example, 0x2497f7b9d7ebb1ee4ee069f26dc396257ca8030ac09d7af8626868dcea193b09
was picked up, but competition data not retrievable since transaction failed. Therefore, by checking status code of 200, we can filter out any failed transactions. Pls let me know if we can restructure how this is done.

Reply:
Very good point. Haven't followed the prior discussion, but would it make sense to actually make the function only operate on a single hash, and then call it multiple times, once per hash, from within the `get_surplus_by_input` function?

Reply:
> Therefore, by checking status code of 200, we can filter out any failed transactions.

And is it possible to distinguish between the entry not being existent in the database vs the api endpoint failing for whatever  reasons?

Reply:
We were sort of doing this before, except for each hash we would compute the surplus directly in a big function. 
What you have suggested is certainly an alternative, but do we need to keep track of failed transactions?

Reply:
to the second comment, yes there is actually. If the entry is not existent, status code = 404. If the comp. data endpoint fails for other reasons, it could be 500 (as seen a few weeks ago)